packages/dashboard/lib/DashboardServer.ts

```diff
@@ -0,0 +1,154 @@
+import {
+  base64ToJson,
+  connectToMessageBusWithRetries,
+  createMessage,
+  DashboardMessageBus,
+  LogMessage,
+  sendAndAwait
+} from "@truffle/dashboard-message-bus";
+import cors from "cors";
+import { Server } from "http";
+import debugModule from "debug";
+import { DashboardServerOptions } from "./types";
+
+export default class DashboardServer {
+  port: number;
+  host: string;
+  rpc: boolean;
+  verbose: boolean;
+  autoOpen: boolean;
+  frontendPath: string;
+
+  private expressApp?: Application;
+  private httpServer?: Server;
+  private messageBus?: DashboardMessageBus;
+  private socket?: WebSocket;
+
+  boundTerminateListener: () => void;
+
+  constructor(options: DashboardServerOptions) {
+    this.port = options.port;
+    this.host = options.host ?? "localhost";
+    this.rpc = options.rpc ?? true;
+    this.verbose = options.verbose ?? false;
+    this.autoOpen = options.autoOpen ?? true;
+    this.frontendPath = path.join(
+      __dirname,
+      ".",
+      "dashboard-frontend",
+      "build"
+    );
+
+    this.boundTerminateListener = () => this.stop();
+  }
+
+  async start() {
+    if (this.httpServer?.listening) return;
+
+    this.messageBus = await this.startMessageBus();
+
+    this.expressApp = express();
+
+    this.expressApp.use(cors());
+    this.expressApp.use(express.json());
+    this.expressApp.use(express.static(this.frontendPath));
+
+    this.expressApp.get("/ports", this.getPorts.bind(this));
+
+    if (this.rpc) {
+      this.socket = await this.connectToMessageBus();
+      this.expressApp.post("/rpc", this.postRpc.bind(this));
+    }
+
+    await new Promise<void>(resolve => {
+      this.httpServer = this.expressApp!.listen(this.port, this.host, () => {
+        if (this.autoOpen) {
+          const host = this.host === "0.0.0.0" ? "localhost" : this.host;
+          open(`http://${host}:${this.port}`);
+        }
+        resolve();
+      });
+    });
+  }
+
+  async stop() {
+    this.messageBus?.off("terminate", this.boundTerminateListener);
+    await this.messageBus?.terminate();
+    this.socket?.terminate();
+    return new Promise<void>(resolve => {
+      this.httpServer?.close(() => resolve());
+    });
+  }
+
+  private getPorts(req: Request, res: Response) {
+    if (!this.messageBus) {
+      throw new Error("Message bus has not been started yet");
+    }
+
+    res.json({
+      dashboardPort: this.port,
+      subscribePort: this.messageBus.subscribePort,
+      publishPort: this.messageBus.publishPort
+    });
+  }
+
+  private postRpc(req: Request, res: Response, next: NextFunction) {
+    if (!this.socket) {
+      throw new Error("Not connected to message bus");
+    }
+
+    const message = createMessage("provider", req.body);
+    sendAndAwait(this.socket, message)
+      .then(response => res.json(response.payload))
+      .catch(next);
+  }
+
+  private async startMessageBus() {
+    const messageBusPublishPort = await getPort({ host: this.host });
```

Comment:
This is not something to hold this PR back but I'm curious, what does passing it the `host` do? I read the docs and I'm still not sure I understand what this does.

Reply:
If there are multiple local addresses, this allows you to specify for which address you want to get an available port.